master/buildbot/worker/libvirt.py

```diff
@@ -33,175 +34,112 @@
     libvirt = None
 
 
-class WorkQueue:
+def handleConnectClose(conn, reason, opaque):
+    opaque.closeConnect()
 
-    """
-    I am a class that turns parallel access into serial access.
-
-    I exist because we want to run libvirt access in threads as we don't
-    trust calls not to block, but under load libvirt doesn't seem to like
-    this kind of threaded use.
-    """
-
-    def __init__(self):
-        self.queue = []
-
-    def _process(self):
-        log.msg("Looking to start a piece of work now...")
-
-        # Is there anything to do?
-        if not self.queue:
-            log.msg("_process called when there is no work")
-            return
-
-        # Peek at the top of the stack - get a function to call and
-        # a deferred to fire when its all over
-        d, next_operation, args, kwargs = self.queue[0]
-
-        # Start doing some work - expects a deferred
-        try:
-            d2 = next_operation(*args, **kwargs)
-        except Exception:
-            d2 = defer.fail()
-
-        # Whenever a piece of work is done, whether it worked or not
-        # call this to schedule the next piece of work
-        @d2.addBoth
-        def _work_done(res):
-            log.msg("Completed a piece of work")
-            self.queue.pop(0)
-            if self.queue:
-                log.msg("Preparing next piece of work")
-                eventually(self._process)
-            return res
-
-        # When the work is done, trigger d
-        d2.chainDeferred(d)
-
-    def execute(self, cb, *args, **kwargs):
-        kickstart_processing = not self.queue
-        d = defer.Deferred()
-        self.queue.append((d, cb, args, kwargs))
-        if kickstart_processing:
-            self._process()
-        return d
-
-    def executeInThread(self, cb, *args, **kwargs):
```

Comment:
I wasn't aware of this code. This is not very useful, If we want serialisaion of calls to libvirt, because it is not threadsafe, the correct solution is to use a threadpool of 1 thread, then twisted will de-facto serialize it.

https://twistedmatrix.com/documents/current/api/twisted.python.threadpool.ThreadPool.html